### PR TITLE
Preserve volume uid and gid through subsequent commands

### LINKED ISSUE #2202: RUN command after VOLUME changes mountpoint owner
**Description**

Any `RUN` command following a `VOLUME` command changes the volume user and group ownership.

**Steps to reproduce the issue:**
1. Build the following two Dockerfiles using `buildah bud`
```dockerfile
FROM centos:8
RUN dnf -y install --setopt=tsflags=nodocs postgresql-server
VOLUME [ "/var/lib/pgsql/data" ]
```

```dockerfile
FROM centos:8
RUN dnf -y install --setopt=tsflags=nodocs postgresql-server
VOLUME [ "/var/lib/pgsql/data" ]
RUN touch /root/test
```
2. Run the resulting images with the following command
```bash
podman run --rm <image> ls -l /var/lib/pgsql
```

**Describe the results you received:**

Image one:
```
$ podman run --rm 5237d257a48dc623f3c1ecb00ccfc0a2eba00daa2895d75339db8da21cec1a14 ls -l /var/lib/pgsql
total 0
drwx------. 1 postgres postgres 0 May 31  2019 backups
drwx------. 1 postgres postgres 0 Mar  3 18:57 data
```
Image two:
```
$ podman run --rm 51a2b48b5a97cd0b05eb942239534595093e57012821befbf2843bba5f68dc2c ls -l /var/lib/pgsql
total 0
drwx------. 1 postgres postgres 0 May 31  2019 backups
drwx------. 1 root     root     0 Mar  3 18:58 data
```
**Describe the results you expected:**
I expected the user and group of the `data` directory to not be changed by running `touch /root/test`

**Output of `rpm -q buildah` or `apt list buildah`:**

```
buildah-1.12.0-2.fc30.x86_64
```

**Output of `buildah version`:**

```
Version:         1.12.0
Go Version:      go1.12.13
Image Spec:      1.0.1-dev
Runtime Spec:    1.0.1-dev
CNI Spec:        0.4.0
libcni Version:  
image Version:   5.0.0
Git Commit:      
Built:           Wed Dec 31 19:00:00 1969
OS/Arch:         linux/amd64
```

**Output of `podman version` if reporting a `podman build` issue:**

```
Version:            1.7.0
RemoteAPI Version:  1
Go Version:         go1.12.14
OS/Arch:            linux/amd64
```

**Output of `cat /etc/*release`:**

```
Fedora release 30 (Thirty)
NAME=Fedora
VERSION="30 (Workstation Edition)"
ID=fedora
VERSION_ID=30
VERSION_CODENAME=""
PLATFORM_ID="platform:f30"
PRETTY_NAME="Fedora 30 (Workstation Edition)"
ANSI_COLOR="0;34"
LOGO=fedora-logo-icon
CPE_NAME="cpe:/o:fedoraproject:fedora:30"
HOME_URL="https://fedoraproject.org/"
DOCUMENTATION_URL="https://docs.fedoraproject.org/en-US/fedora/f30/system-administrators-guide/"
SUPPORT_URL="https://fedoraproject.org/wiki/Communicating_and_getting_help"
BUG_REPORT_URL="https://bugzilla.redhat.com/"
REDHAT_BUGZILLA_PRODUCT="Fedora"
REDHAT_BUGZILLA_PRODUCT_VERSION=30
REDHAT_SUPPORT_PRODUCT="Fedora"
REDHAT_SUPPORT_PRODUCT_VERSION=30
PRIVACY_POLICY_URL="https://fedoraproject.org/wiki/Legal:PrivacyPolicy"
VARIANT="Workstation Edition"
VARIANT_ID=workstation
Fedora release 30 (Thirty)
Fedora release 30 (Thirty)
```

**Output of `uname -a`:**

```
Linux localhost.localdomain 5.4.19-100.fc30.x86_64 #1 SMP Tue Feb 11 22:27:11 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
```

**Output of `cat /etc/containers/storage.conf`:**

```
# This file is is the configuration file for all tools
# that use the containers/storage library.
# See man 5 containers-storage.conf for more information
# The "container storage" table contains all of the server options.
[storage]

# Default Storage Driver
driver = "overlay"

# Temporary storage location
runroot = "/var/run/containers/storage"

# Primary Read/Write location of container storage
graphroot = "/var/lib/containers/storage"

[storage.options]
# Storage options to be passed to underlying storage drivers

# AdditionalImageStores is used to pass paths to additional Read/Only image stores
# Must be comma separated list.
additionalimagestores = [
]

# Size is used to set a maximum size of the container image.  Only supported by
# certain container storage drivers.
size = ""

# Path to an helper program to use for mounting the file system instead of mounting it
# directly.
#mount_program = "/usr/bin/fuse-overlayfs"

# OverrideKernelCheck tells the driver to ignore kernel checks based on kernel version
override_kernel_check = "true"

# mountopt specifies comma separated list of extra mount options
mountopt = "nodev,metacopy=on"

# Remap-UIDs/GIDs is the mapping from UIDs/GIDs as they should appear inside of
# a container, to UIDs/GIDs as they should appear outside of the container, and
# the length of the range of UIDs/GIDs.  Additional mapped sets can be listed
# and will be heeded by libraries, but there are limits to the number of
# mappings which the kernel will allow when you later attempt to run a
# container.
#
# remap-uids = 0:1668442479:65536
# remap-gids = 0:1668442479:65536

# Remap-User/Group is a name which can be used to look up one or more UID/GID
# ranges in the /etc/subuid or /etc/subgid file.  Mappings are set up starting
# with an in-container ID of 0 and the a host-level ID taken from the lowest
# range that matches the specified name, and using the length of that range.
# Additional ranges are then assigned, using the ranges which specify the
# lowest host-level IDs first, to the lowest not-yet-mapped container-level ID,
# until all of the entries have been used for maps.
#
# remap-user = "storage"
# remap-group = "storage"

[storage.options.thinpool]
# Storage Options for thinpool

# autoextend_percent determines the amount by which pool needs to be
# grown. This is specified in terms of % of pool size. So a value of 20 means
# that when threshold is hit, pool will be grown by 20% of existing
# pool size.
# autoextend_percent = "20"

# autoextend_threshold determines the pool extension threshold in terms
# of percentage of pool size. For example, if threshold is 60, that means when
# pool is 60% full, threshold has been hit.
# autoextend_threshold = "80"

# basesize specifies the size to use when creating the base device, which
# limits the size of images and containers.
# basesize = "10G"

# blocksize specifies a custom blocksize to use for the thin pool.
# blocksize="64k"

# directlvm_device specifies a custom block storage device to use for the
# thin pool. Required if you setup devicemapper.
# directlvm_device = ""

# directlvm_device_force wipes device even if device already has a filesystem.
# directlvm_device_force = "True"

# fs specifies the filesystem type to use for the base device.
# fs="xfs"

# log_level sets the log level of devicemapper.
# 0: LogLevelSuppress 0 (Default)
# 2: LogLevelFatal
# 3: LogLevelErr
# 4: LogLevelWarn
# 5: LogLevelNotice
# 6: LogLevelInfo
# 7: LogLevelDebug
# log_level = "7"

# min_free_space specifies the min free space percent in a thin pool require for
# new device creation to succeed. Valid values are from 0% - 99%.
# Value 0% disables
# min_free_space = "10%"

# mkfsarg specifies extra mkfs arguments to be used when creating the base.
# device.
# mkfsarg = ""

# use_deferred_removal marks devicemapper block device for deferred removal.
# If the thinpool is in use when the driver attempts to remove it, the driver 
# tells the kernel to remove it as soon as possible. Note this does not free
# up the disk space, use deferred deletion to fully remove the thinpool.
# use_deferred_removal = "True"

# use_deferred_deletion marks thinpool device for deferred deletion.
# If the device is busy when the driver attempts to delete it, the driver
# will attempt to delete device every 30 seconds until successful.
# If the program using the driver exits, the driver will continue attempting
# to cleanup the next time the driver is used. Deferred deletion permanently
# deletes the device and all data stored in device will be lost.
# use_deferred_deletion = "True"

# xfs_nospace_max_retries specifies the maximum number of retries XFS should
# attempt to complete IO when ENOSPC (no space) error is returned by
# underlying storage device.
# xfs_nospace_max_retries = "0"

# If specified, use OSTree to deduplicate files with the overlay backend
ostree_repo = ""

# Set to skip a PRIVATE bind mount on the storage home directory.  Only supported by
# certain container storage drivers
skip_mount_home = "false"
```

### PR DESCRIPTION
**Pull request by carbonin**
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

This PR pulls the UID and GID fields from the stored file stat when reconstructing a volume after subsequent commands

#### How to verify it
Build the following Dockerfile:
```Dockerfile
FROM centos:8
RUN dnf -y install --setopt=tsflags=nodocs postgresql-server
VOLUME [ "/var/lib/pgsql/data" ]
RUN touch /root/test
```

Observe the output of the following command:
```bash
$ podman run --rm <image> ls -l /var/lib/pgsql
total 0
drwx------. 1 postgres postgres 0 May 31  2019 backups
drwx------. 1 postgres postgres 0 Mar 10 23:37 data
```

#### Which issue(s) this PR fixes:

Fixes #2202

#### Does this PR introduce a user-facing change?
No

```release-note
None
```

### CODE REVIEW

**Comment by carbonin** [comment]
I also put together a simpler Dockerfile that would make a good test case, but this just fails to build as a failure. I'm not sure what the `bats` test would look like. May give that a shot tomorrow if I get the chance.

```dockerfile
FROM alpine
RUN adduser -D -H testuser && addgroup testgroup
RUN mkdir -p /vol/subvol
RUN chown testuser:testgroup /vol/subvol
VOLUME /vol/subvol
RUN touch /test
RUN MY_USER=$(stat -c "%U" /vol/subvol) && [[ "$MY_USER" == "testuser" ]]
RUN MY_GROUP=$(stat -c "%G" /vol/subvol) && [[ "$MY_GROUP" == "testgroup" ]]
```

**Comment by carbonin** [comment]
Also I'm getting some permission denied issues running the tests locally, should those need to be run as root?

**Comment by rhatdan** [comment]
This is what I get when running your Dockerfile, as non root. 
$ rpm -q podman
podman-1.8.1-0.7.rc4.fc32.x86_64
$ podman info | grep Buildah
BuildahVersion: 1.14.2

```
$ podman build /tmp
STEP 1: FROM alpine
STEP 2: RUN adduser -D -H testuser && addgroup testgroup
--> fd0d785bdc1
STEP 3: RUN mkdir -p /vol/subvol
--> 8b51f233d8e
STEP 4: RUN chown testuser:testgroup /vol/subvol
--> 943b4d301d9
STEP 5: VOLUME /vol/subvol
--> 309203291ef
STEP 6: RUN touch /test
--> e30bab86456
STEP 7: RUN MY_USER=$(stat -c "%U" /vol/subvol) && [[ "$MY_USER" == "testuser" ]]
Error: error building at STEP "RUN MY_USER=$(stat -c "%U" /vol/subvol) && [[ "$MY_USER" == "testuser" ]]": error while running runtime: exit status 1
```

**Comment by carbonin** [comment]
Yes, that was the test case. With my patch that build succeeds @rhatdan

**Comment by carbonin** [comment]
Sorry, I think I misunderstood your concern. The permission denied issues I was seeing were from running the whole `bud.bats` suite. That Dockerfile fails because of the issue I'm trying to fix, not because of a permissions problem.

**Comment by rhatdan** [comment]
Now I understand.
LGTM

**Comment by rhatdan** [comment]
@nalind @TomSweeneyRedHat @vrothberg @giuseppe @QiWang19 PTAL

**Comment by rhatdan** [comment]
@carbonin Take your test Dockerfile and make it a test case to make sure we don't digress.
Excellent work.

File: tests/bud.bats [hunk 1]
```diff
@@ -599,6 +599,18 @@ load helpers
...
+@test "bud-volume-ownership" {
+  # This Dockerfile needs us to be able to handle a working RUN instruction.
+  skip_if_no_runtime
+
+  _prefetch alpine
+  target=volume-image
+  run_buildah bud --signature-policy ${TESTSDIR}/policy.json -t ${target} ${TESTSDIR}/bud/volume-ownership
+  run_buildah from --quiet --signature-policy ${TESTSDIR}/policy.json ${target}
+  cid=$output
+  run_buildah run $cid test-ownership
```
Comment (carbonin) [inline tests/bud.bats]: I didn't use `buildah mount` here because it changes the ownership of the volume which is what I'm trying to test.

**Comment by nalind** [comment]
LGTM

**Comment by TomSweeneyRedHat** [comment]
LGTM
and TYVM @carbonin !

**Pull request merged**